walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,224 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Set;
+
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.File;
+import java.io.UnsupportedEncodingException;
+
 
 public final class FindMeetingQuery {
+
+  /**
+   * Returns an ArrayList of TimeRanges when the MeetingRequest can happen
+   * If it is possible for all optional attendees to participate, returns those ranges
+   * Else returns ranges when mandatory attendees can participate
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // Create a Set with both optional and mandatory attendees
+    Set<String> requestAllAttendees = new HashSet<String>();
+    requestAllAttendees.addAll(request.getAttendees());
+    requestAllAttendees.addAll(request.getOptionalAttendees());
+
+    // Try to include both optional and mandatory attendees
+    Collection<TimeRange> answer = 
+      queryMandatories(events, new MeetingRequest(requestAllAttendees, request.getDuration()));
+
+    // If there is no suitable range, include only the mandatory attendees from request
+    if(answer.isEmpty()) {
+      answer = queryMandatories(events, request);
+    }
+
+    return answer;
+  }
+
+  /**
+   * Returns an ArrayList of TimeRanges when the MeetingRequest can happen
+   * Takes into consideration only mandatory attendees of the MeetingRequest
+   *
+   * Algorithm: 
+   * Each event has a starting and an ending point, put them all in an array
+   * Sort the array to process the points in ascending order
+   * The meeting will always start when an event ends so try all those cases
+   * The meeting will always end when an event starts
+   * Considering we are processing the starting points in ascending order, the meeting 
+   * for a starting point will always end after or at the same time with the starting point before
+   *
+   * Complexity: O(nlogn + m)
+   * n = number of events
+   * m = total number of attendees in all the events
+   * nlogn : we sort the events
+   * m : we go at most four times (two times with startIndex, two times with endIndex)
+   * through each event and each time we check which attendees of the event are in the MeetingRequest
+   */
+  private Collection<TimeRange> queryMandatories(Collection<Event> events, MeetingRequest request) {
+    ArrayList<TimeRange> answer = new ArrayList<>();
```

Comment:
It's better to use a proper named variable for this. Especially because the function is of non-trivial size. How about 'meetings' or smth like that?

Reply:
Right, thanks! I changed it to availableRanges, I think it sounds better